### PR TITLE
feat(cpu): generate c bindings

### PR DESCRIPTION
**Pull request by sarah-ek**
### Description
adds a `build.rs` file that generates the c bindings for concrete-cpu

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [ ] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [ ] The draft release description has been updated
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: concrete-cpu/src/types.rs [hunk 1]
```diff
@@ -1,11 +1,9 @@
 use concrete_core::backends::fft::private::math::fft::Fft as FftImpl;
 
-#[repr(transparent)]
```
Comment (mayeul-zama) [inline concrete-cpu/src/types.rs:3]: Did `#[repr(transparent)]` prevent generation of C type definitions?

**Review by mayeul-zama — COMMENTED** [review]
I think it would be good to have the generated header file be part of the repository (and checked by the CI to be equal to the one generated) so that a API change can be seen from PRs (or from the CI if the file has not been updated)

File: concrete-cpu/src/types.rs (hunk 1 above)
Comment (sarah-ek) [reply concrete-cpu/src/types.rs:3]: i'm not sure why, but it caused the build script to panic

**Comment by sarah-ek** [comment]
@mayeul-zama i'm not sure how to do what you asked. the part about making the CI check if the generated header file is the same as the one that's committed

**Comment by agnesLeroy** [comment]
> @mayeul-zama i'm not sure how to do what you asked. the part about making the CI check if the generated header file is the same as the one that's committed

@sarah-ek you have an example in .github/workflows/lint.yml at line 56 of a CI check that fails if there's a diff in the repo, maybe it can help you?

**Comment by mayeul-zama** [comment]
> > @mayeul-zama i'm not sure how to do what you asked. the part about making the CI check if the generated header file is the same as the one that's committed
> 
> @sarah-ek you have an example in concrete-core/.github/workflows/lint.yml at line 56 of a CI check that fails if there's a diff in the repo, maybe it can help you?

In the optimizer we also have a similar check with c++ files generated by cxx
https://github.com/zama-ai/concrete-optimizer/blob/0d7cb97e7ebe191b52cd833fc25b218f66ad4c3b/concrete-optimizer-cpp/Makefile#L46 used there https://github.com/zama-ai/concrete-optimizer/blob/0d7cb97e7ebe191b52cd833fc25b218f66ad4c3b/.github/workflows/optimizer.yml#L39

File: .github/workflows/cpu.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install Rust for tasks
+        run: |
+          make install_tasks_rust_toolchain
+      - name: Run cbindgen
+        run: |
+          cargo build --package concrete-cpu
+      - name: Check diff
+        run: git diff --exit-code
```
Comment (mayeul-zama) [inline .github/workflows/cpu.yml:20]: Did you check that this job fails in the CI if `concrete-cpu.h` is wrong?
I think it's better to make sure it's working correctly

Comment (sarah-ek) [reply .github/workflows/cpu.yml:20]: how should i check? do i open a separate PR or just push to this one?

Comment (sarah-ek) [reply .github/workflows/cpu.yml:20]: yeah, it seems to work

**Comment by agnesLeroy** [comment]
@sarah-ek do you have the rights to merge on the repo?

**Comment by sarah-ek** [comment]
@agnesLeroy no, i don't

**Comment by agnesLeroy** [comment]
> @agnesLeroy no, i don't

I think you can ask @soonum for this, it's probably going to be more practical for you to work on concrete-cpu. I'm merging this PR in the meanwhile.

**Comment by soonum** [comment]
Weird @sarah-ek, I see that prior to become an admin for concrete-cpu, you had the `write` role. I believe that write access to a repo allows to perform the merge.

**Pull request merged**